File: kt/main/kotlin/pushtofunction/fs/boot/BarBfBoot.kt
/*
 * Copyright © 2022 Paulo Villela. All rights reserved.
 * Use of this source code is governed by the MIT license
 * that can be found in the LICENSE file.
 */

package pushtofunction.fs.boot

import pushtofunction.fs.BarBfCfgInfo
import pushtofunction.fs.BarBfCfgSrc
import pushtofunction.fs.BarBfT
import pushtofunction.fs.barBfC
import pushtofunction.config.AppCfgInfo
import pushtovar.fwk.liftToNullary

private fun barBfCfgAdapter0(appCfg: AppCfgInfo): BarBfCfgInfo {
	return BarBfCfgInfo(
		z = appCfg.y,
	)
}

var barBfCfgAdapter = liftToNullary(::barBfCfgAdapter0)

fun barBfBoot(appCfg: (() -> AppCfgInfo)?): BarBfT {
	val barBfCfgSrc = BarBfCfgSrc(
		get = barBfCfgAdapter(appCfg)
	)
	return barBfC(barBfCfgSrc)
}

File: rs/pushtofunction/src/fs/boot/bar_bf_boot.rs
use crate::fs::{bar_bf_c, BarBfCfgDeps, BarBfT};
use common::config::AppCfgInfo;
use common::fs_data::BarBfCfgInfo;
use common::fwk::RefreshMode;
use std::sync::Arc;

fn bar_bf_cfg_adapter(app_cfg: &AppCfgInfo) -> BarBfCfgInfo {
    BarBfCfgInfo {
        u: app_cfg.y,
        v: app_cfg.x.clone(),
    }
}

pub fn bar_bf_boot(app_cfg: fn() -> Arc<AppCfgInfo>, refresh_mode: RefreshMode) -> BarBfT {
    let bar_bf_cfg_deps =
        BarBfCfgDeps::new_boxed_with_cfg_adapter(app_cfg, bar_bf_cfg_adapter, refresh_mode, ());
    bar_bf_c(bar_bf_cfg_deps)
}
